File: internal/handler/automation.go
package handler

import (
	"strconv"

	"github.com/aiops/aiops-platform/internal/automation"
	"github.com/aiops/aiops-platform/pkg/response"
	"github.com/gin-gonic/gin"
)

// AutomationHandler 处理自动化运维请求。
type AutomationHandler struct {
	Engine *automation.Engine
}

// GetPodLogs 处理 GET /api/v1/automation/pods/:pod/logs
func (h *AutomationHandler) GetPodLogs(c *gin.Context) {
	if h.Engine == nil {
		response.Internal(c, "自动化引擎未初始化")
		return
	}

	pod := c.Param("pod")
	cluster := c.DefaultQuery("cluster", "")
	namespace := c.DefaultQuery("namespace", "default")
	container := c.Query("container")
	tail := int64(100)
	if t := c.Query("tail"); t != "" {
		if v, err := strconv.ParseInt(t, 10, 64); err == nil && v > 0 {
			tail = v
		}
	}

	logs, err := h.Engine.GetPodLogs(c.Request.Context(), cluster, namespace, pod, container, tail)
	if err != nil {
		response.Internal(c, "获取 Pod 日志失败: "+err.Error())
		return
	}

	response.OK(c, gin.H{"logs": logs})
}

// GetPodEvents 处理 GET /api/v1/automation/pods/:pod/events
func (h *AutomationHandler) GetPodEvents(c *gin.Context) {
	if h.Engine == nil {
		response.Internal(c, "自动化引擎未初始化")
		return
	}

	pod := c.Param("pod")
	cluster := c.DefaultQuery("cluster", "")
	namespace := c.DefaultQuery("namespace", "default")

	events, err := h.Engine.GetPodEvents(c.Request.Context(), cluster, namespace, pod)
	if err != nil {
		response.Internal(c, "获取 Pod Event 失败: "+err.Error())
		return
	}

	response.OK(c, events)
}

// restartRequest 重启 Pod 请求体。
type restartRequest struct {
	Cluster   string `json:"cluster"`
	Namespace string `json:"namespace"`
	Confirm   bool   `json:"confirm"`
}

// RestartPod 处理 POST /api/v1/automation/pods/:pod/restart
func (h *AutomationHandler) RestartPod(c *gin.Context) {
	if h.Engine == nil {
		response.Internal(c, "自动化引擎未初始化")
		return
	}

	pod := c.Param("pod")
	var req restartRequest
	if err := c.ShouldBindJSON(&req); err != nil {
		response.BadRequest(c, "请求体格式错误: "+err.Error())
		return
	}
	if req.Namespace == "" {
		req.Namespace = "default"
	}

	if err := h.Engine.RestartPod(c.Request.Context(), req.Cluster, req.Namespace, pod, req.Confirm); err != nil {
		response.BadRequest(c, err.Error())
		return
	}

	response.OK(c, gin.H{"message": "Pod 重启指令已发送", "pod": pod, "namespace": req.Namespace})
}

// scaleRequest 扩容请求体。
type scaleRequest struct {
	Cluster   string `json:"cluster"`
	Namespace string `json:"namespace"`
	Replicas  int32  `json:"replicas"`
	Confirm   bool   `json:"confirm"`
}

// ScaleDeployment 处理 POST /api/v1/automation/deployments/:name/scale
func (h *AutomationHandler) ScaleDeployment(c *gin.Context) {
	if h.Engine == nil {
		response.Internal(c, "自动化引擎未初始化")
		return
	}

	name := c.Param("name")
	var req scaleRequest
	if err := c.ShouldBindJSON(&req); err != nil {
		response.BadRequest(c, "请求体格式错误: "+err.Error())
		return
	}
	if req.Namespace == "" {
		req.Namespace = "default"
	}
	if req.Replicas < 0 {
		response.BadRequest(c, "replicas 不能为负数")
		return
	}

	if err := h.Engine.ScaleDeployment(c.Request.Context(), req.Cluster, req.Namespace, name, req.Replicas, req.Confirm); err != nil {
		response.BadRequest(c, err.Error())
		return
	}

	response.OK(c, gin.H{"message": "Deployment 扩容指令已发送", "deployment": name, "replicas": req.Replicas})
}

File: frontend/src/api/automation.ts
import request from './client'

export interface AutomationAction {
  id: number
  incident_id?: number
  user_id: number
  action_type: string
  target_type: string
  target_name: string
  cluster: string
  connection_id?: number
  namespace: string
  parameters: string
  reason: string
  risk: string
  status: string
  approved_by?: number
  approved_at?: string
  rejected_by?: number
  rejected_at?: string
  reject_reason?: string
  created_at: string
  updated_at: string
}

export interface DryRunResult {
  action_type: string
  target: string
  current_state: string
  expected_operation: string
  potential_impact: string
  safe: boolean
}

export interface ExecutionResult {
  success: boolean
  message: string
  data?: any
  error?: string
}

export interface ActionExecution {
  id: number
  action_id: number
  executor: string
  started_at: string
  finished_at?: string
  duration_ms: number
  status: string
  result_json?: string
  error?: string
  created_at: string
}

export const automationApi = {
  create: (data: {
    incident_id?: number
    action_type: string
    target_type: string
    target_name: string
    cluster: string
    connection_id?: number
    namespace?: string
    parameters?: Record<string, any>
    reason: string
    risk?: string
  }) => request.post<any, AutomationAction>('/api/v1/actions', data),

  list: (params?: {
    status?: string
    risk?: string
    action_type?: string
    incident_id?: number
    cluster?: string
    page?: number
    page_size?: number
  }) => request.get<any, { items: AutomationAction[]; total: number; page: number; page_size: number }>('/api/v1/actions', { params }),

  pendingApproval: (params?: { page?: number; page_size?: number }) =>
    request.get<any, { items: AutomationAction[]; total: number; page: number; page_size: number }>('/api/v1/actions/pending-approval', { params }),

  createFromIncident: (incidentId: number, data: {
    action_type: string
    target_type: string
    target_name: string
    cluster: string
    connection_id?: number
    namespace?: string
    parameters?: Record<string, any>
    reason: string
    risk?: string
  }) => request.post<any, AutomationAction>(`/api/v1/incidents/${incidentId}/actions`, data),

  get: (id: number) => request.get<any, AutomationAction>(`/api/v1/actions/${id}`),

  approve: (id: number) => request.post<any, AutomationAction>(`/api/v1/actions/${id}/approve`),

  reject: (id: number, reason: string) => request.post<any, AutomationAction>(`/api/v1/actions/${id}/reject`, { reason }),

  dryRun: (id: number) => request.post<any, DryRunResult>(`/api/v1/actions/${id}/dry-run`),

  execute: (id: number) => request.post<any, ExecutionResult>(`/api/v1/actions/${id}/execute`),

  cancel: (id: number) => request.post<any, AutomationAction>(`/api/v1/actions/${id}/cancel`),

  executions: (id: number) => request.get<any, ActionExecution[]>(`/api/v1/actions/${id}/executions`),

  audit: (params?: { action_id?: number; incident_id?: number; user_id?: number; page?: number; page_size?: number }) =>
    request.get<any, { items: any[]; total: number }>('/api/v1/automation/audit', { params }),

  // 旧版直接操作 API（Pod Detail / Deployment Detail 使用）
  podLogs: (pod: string, params?: { cluster?: string; namespace?: string; container?: string; tail_lines?: number; tail?: number }) =>
    request.get<any, string>(`/api/v1/automation/pods/${pod}/logs`, { params }),

  podEvents: (pod: string, params?: { cluster?: string; namespace?: string }) =>
    request.get<any, any[]>(`/api/v1/automation/pods/${pod}/events`, { params }),

  restartPod: (pod: string, data: { cluster: string; namespace: string; confirm: boolean }) =>
    request.post<any, any>(`/api/v1/automation/pods/${pod}/restart`, data),

  scaleDeployment: (name: string, data: { cluster: string; namespace: string; replicas: number; confirm: boolean }) =>
    request.post<any, any>(`/api/v1/automation/deployments/${name}/scale`, data),
}
